Automattic/wp-calypso
# Allow all primary My Home cards to be rendered in secondary

Related to #83465

## Proposed Changes

* This PR refactors the code for primary and secondary cards in My Home so we can more clearly identify primary-only cards, and allow all of those cards to be rendered in the secondary area.
* The work here is a precursor to some related work where we want to shift My Home to a full two-column layout and show the Site Preview block at the top of the right sidebar, where all cards that were previously in the full-width primary section will be shown in the secondary section next to the sidebar.

  - Actually testing that this works depends on D127655-code, which updates the back end to rearrange the layout so we don't show anything in the primary area, but shift any cards in that location to the top of the secondary area (which is where this PR is needed!)

## Testing Instructions


There are two things we need to test with this PR:
1. That the code works with the current layouts, and
2. That the code works with the new two-column layouts we want to use

### Testing the current layout

* Run this branch via Calypso.live as per [this comment](https://github.com/Automattic/wp-calypso/pull/83828#issuecomment-1791512176), or locally
* For sites in various states, visit My Home for the site, and verify that the UI renders as expected. Valid states we should check include the following:
  - a site with an expired plan
  - a still-Atomic site with an expired plan
  - a site that was Atomic and has been reverted after the plan expired
  - a site with a Launchpad in Customer Home
  - a site where we're showing the site setup options
  - a launched site where the user has completed or skipped the site setup steps, and we're showing general promotional cards in the primary area

### Testing the two-column layout

* Apply D127655-code to your WordPress.com sandbox
* Ensure that requests to `public-api.wordpress.com` are being routed to your sandbox
* Ensure that the following PHP code snippet is running on your sandbox:
```php
add_filter( 'wpcom_my_home_force_two_column_layout', '__return_true' );
```
* Run this branch via Calypso.live as per [this comment](https://github.com/Automattic/wp-calypso/pull/83828#issuecomment-1791512176), or locally
* For sites in various states, visit My Home for the site, and verify that UI renders with two vertical columns, and the right column has the site preview in the first/top position. Note that many of the cards don't look great -- we need to address those cases before we can deploy the proposed layout changes. Valid states we should check include the following:
  - a site with an expired plan
  - a still-Atomic site with an expired plan
  - a site that was Atomic and has been reverted after the plan expired
  - a site with a Launchpad in Customer Home
  - a site where we're showing the site setup options
  - a launched site where the user has completed or skipped the site setup steps, and we're showing general promotional cards in the primary area

## Pre-merge Checklist



- [N/A] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [N/A] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [N/A] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
